계산기 [Step 2] som

jsa0224 opened the pull request:
안녕하세요, 태태! (@uuu1101)
Step 2 구현이 끝나서 PR 올립니다!

# 📍UML(Step 2에서 제공)
![](https://i.imgur.com/wPVOKCQ.png)


# 📍Content
- `Operator`
  - `calculate(lhs:rhs:)`: rawValue를 통해 사칙연산이 가능하다.
  - `add(lhs:rhs:)`: 매개변수로 들어온 값을 더한다.
  - `subtract(lhs:rhs:)`: 매개변수로 들어온 값을 뺀다.
  - `divide(lhs:rhs:)`: 매개변수로 들어온 값을 나눈다.
  - `multiply(lhs:rhs:)`: 매개변수로 들어온 값을 곱한다. 
 
- `Formula`
  - `result()`: 사칙연산을 실행하고 계산의 결과를 반환한다. 

- `ExpressionParser`
  - `parse(from:)`: `String` 형태로 받아온 연산식을 피연산자는 `Double`로, 연산자는 `Operator` 타입으로 `Formula`에게 반환한다.
  - `componentsByOperators(from:)`: 연산자를 기준으로 문자열(연산식)을 나눈 다음, 문자열 배열 형태로 반환한다.


# 📍고민했던 부분
- **inputStack이 nil일 경우, deQueue 메서드가 실행될 때**
inputStack이 nil일 경우, removeLast()를 하지 못 해 컴파일러 에러가 일어나게 됩니다. 그래서 처음에는 inputStack이 nil일 경우, nil을 반환하는 코드를 추가로 구현했습니다.
```swift
mutating func deQueue() -> T? {
    if inputStack.isEmpty {
        return nil
    }
        
    if outputStack.isEmpty {
        outputStack = inputStack.reversed()
        inputStack.removeAll()
    }
    return outputStack.removeLast()
}
```
하지만 이렇게 구현하니 테스트 코드를 돌릴 때, 아래와 같은 오류가 일어났습니다.
![](https://i.imgur.com/RpyNLR9.png)

inputStack이 빈 배열로 deQueue 메서드로 실행되기 때문에 nil 상태가 되어, nil이 리턴되는 popLast()로 다시 수정했습니다. 

- **고차함수**
고차함수를 쓰는 것이 익숙하지 않아서, 조금 어려웠습니다. 개념을 모른다기 보다는 많이 사용해봐야 익숙해질 거 같습니다. 

- **Extension 소스 파일**
Extension 타입마다 파일을 따로 분리할까 고민했지만, "확장"이라는 공통점으로 한 파일에 구현했습니다. 

# 📍조언 받고 싶은 부분
```swift
extension String {
    func split(with target: Character) -> [String] {
        return self.components(separatedBy: String(target))
    }
}
```
제가 원래는 해당 `extension String`의 `split(with:)`를 `ExpressionParser`안에서 썼었는데, 테스트 코드에서 테스트할 때 오류가 나서 해당 메서드를 쓰지 않고 구현했습니다. 지금 코드에서 해당 메서드를 구현할 방법이 있을까요?

uuu1101 commented on Calculator/Calculator/ExpressionParser.swift:
```diff
@@ -0,0 +1,28 @@
+enum ExpressionParser {
+    static func parse(from input: String) -> Formula {
+        let divided = ExpressionParser.componentsByOperators(from: input)
+        let operands = divided.compactMap { Double($0) }
```
`compactMap`은 어떤 메서드 였나요?!

jsa0224 replied:
`compactMap`은 컨테이너의 각 요소를 조건을 지정하여, 호출할 때 `nil`이 아닌 배열을 반환합니다! `map`과 다른 점은 `옵셔널 바인딩`을 한다는 것인데, 그런 의미로 `map`이 아닌 `compactMap`을 사용했습니다.

uuu1101 commented on Calculator/Calculator/Extension.swift:
```diff
@@ -0,0 +1,18 @@
+    }
+}
+
+extension Double {
+    
+}
```
해당 `extension`은 필요한걸까요?

jsa0224 replied:
저도 UML 다이어그램에 있던 내용이라 구현하긴 했지만 아직까지는 해당 `Extension`이 필요한 이유를 모르겠습니다🧐

uuu1101 reviewed (changes requested):
안녕하세요 som! 
주어진 UML을 따라서 잘 구현하셨네요 ! 👍 
커밋 단위도 아주 좋습니다!
현재 하나의 디렉토리에 코드 파일들이 들어있는데, 이 파일들을 폴더링 해보셔도 좋을 것 같아요.
나중에 프로젝트 규모가 커지고 코드 파일들이 많아지면 폴더링하시면 관리하기 편하실거에요 :)
이와 관련해서 의존모둠 분들과 대화를 나눠보아도 좋을 것 같습니다.

저번에 제가 드렸던 질문에 대해서 잘 답변해주셨네요 👍 

- 고차함수 사용하는 것은 참 어려운일이에요, 저도 처음에 사용할때 무척 어려웠거든요..
고차함수에 대한 개념을 알고 계신다니 이 부분은 사용하다보면 익숙해지실거에요, 처음에는 고차함수 축약형을 지양하시고 조금 익숙해지면 그때 축약형을 사용해보시는 것도 하나의 방법입니다 !

- Extension 소스 파일 폴더링에 대해서 
하나의 공통점을 가지고 하나로 묶은점은 정말 좋네요 ! 👍 
하나의 Extension 폴더에 Extension+String.swift 처럼 네이밍을 해서 여러 파일을 하나의 폴더로 관리하는 방법도 있습니다 ! 

- 테스트 코드에서 어떠한 오류가 발생하였을까요?
- UML 에서 '+' 와 '-' 는 어떠한 의미를 가졌었나요?
- UML 에서 밑줄은 어떤 의미를 가졌었나요?

- FormulaTests 와 ExpressionParserTests를 실행하면 실행되지 않습니다.
> 아래 이미지처럼 나오는데 어떤 에러이고 어떻게 해결 할 수 있을까요?! 👀
![스크린샷 2022-09-26 오후 8 21 24](https://user-images.githubusercontent.com/49808034/192263990-7ae47164-7bcd-4885-8ef9-41a8726b91d0.png)

공부를 하시다가, 프로젝트를 진행하시면서 어려운 부분이나 궁금하신 부분이 있다면 언제든 DM 주세요!

jsa0224 commented:
안녕하세요, 태태! 좋은 질문들을 주셔서 제가 생각해보느라 답변이 늦어졌습니다..

- **Extension 소스 파일링**
네, 저도 한 파일에 구현하긴 했지만, 코드 양이 방대해지면 태태가 알려주신 방법으로 폴더링해보겠습니다! 

- **테스트 코드에서 생긴 오류**
ExpressionParser의 componentsByOperators 메서드를 split(with:)로 구현했을 때, 
```swift
var arithmeticExpressionArray = [input]
var tempArray: [String] = []

Operator.allCases.forEach { oneOfOperator in
    let character = oneOfOperator.rawValue
    for arithmeticExpression in arithmeticExpressionArray {
            tempArray += arithmeticExpression.split(with: character)
            arithmeticExpressionArray = tempArray
            tempArray.removeAll()
        }
    }

return arithmeticExpressionArray
```
```swift
func split(with target: Character) -> [String] {
    let splited = String(target)
    let replaced = self.replacingOccurrences(of: splited, with: " \(splited) ")
    let separator = replaced.split(separator: " ").map { String($0) }

    return separator
```
ExpressionParserTest에서 테스트를 실행하니
```swift
func test_10_더하기_20_곱하기_30_나누기_3은_300인가() throws {
    let string = "10+20*30/3"
    var parseTest = ExpressionParser.parse(from: string)
    let result = try parseTest.result()
        
    XCTAssertEqual(result, 300.0)
    }
```
테스트가 실패한다는 결과가 나와, 연산자가 잘 전달되지 않는다는 느낌을 받아, PR을 보낼 당시의 코드로 수정했습니다. 현재는 태태께 디엠으로 질문 드린 것처럼 다른 코드로 다시 구현해서 내용이 변경되었을 것입니다!

- **UML에서 '+'와 '-'는 어떠한 의미를 가졌나요?**
접근제어자의 의미를 가지고 있습니다. '+'는 `public`, '-'는 `private`으로 표현합니다. 
 
- **UML에서 밑줄은 어떤 의미를 가졌었나요?**
`Static Attribute / Operation`인 경우 밑줄을 그어 표기합니다. 이 프로젝트의 경우 `타입 메서드`를 표현하려고 쓰인 것 같습니다. 

- **리팩토링**
0으로 나눴을 때의 예외 처리와 음수 연산 관련 코드를 추가, 폴더링 수정하여 리팩토링 보냅니다!

uuu1101 commented:
split 을 하였을때 연산자가 전달이 제대로 안된다고 하셨는데, 디버깅을 통해서 확인해보는건 어떨까요?!
